lib/prism/translation/parser/compiler.rb

```diff
@@ -1607,11 +1607,7 @@ def visit_when_node(node)
           builder.when(
             token(node.keyword_loc),
             visit_all(node.conditions),
-            if node.then_keyword_loc
-              token(node.then_keyword_loc)
-            else
-              srange_find(node.conditions.last.location.end_offset, node.statements&.location&.start_offset || (node.conditions.last.location.end_offset + 1), [";"])
-            end,
+            srange_find(node.conditions.last.location.end_offset, node.statements&.location&.start_offset || node.statements&.then_keyword_loc, [";", "then"]),
```

Comment:
Note: This is the same solution as the following.
https://github.com/ruby/prism/blob/v0.24.0/lib/prism/translation/parser/compiler.rb#L828